dsl/flatten.cc

```diff
@@ -0,0 +1,304 @@
+#include "dsl/flatten.h"
+#include "ast/Helpers.h"
+#include "ast/ast.h"
+#include "ast/treemap/treemap.h"
+#include "core/core.h"
+
+#include <utility>
+
+using namespace std;
+
+// This DSL pass flattens nested methods, so that once we've reached a non-definition AST node (i.e. not a ClassDef or a
+// MethodDef) then we know that there are no MethodDefs lurking deeper in the tree. In order to work correctly, this
+// also needs to move some non-method-def things as well, specifically `sig`s and sends for method visibility
+// (e.g. `private` and the like), and it also updates the static-ness of some MethodDefs based on where they have
+// appeared in a nested context.
+//
+// So, a file like the following
+//
+// class A
+//   sig{void}
+//   private def foo
+//     sig{void}
+//     def self.bar; end
+//   end
+// end
+//
+// will morally be transformed into the following
+//
+// class A
+//   sig{void}
+//   private def foo; :bar; end
+//   sig{void}
+//   def bar; end   # notice the lack of `self.` here
+// end
+//
+// So no nested methods exist any longer, and additionally, the nested method `bar` has had the `self.` qualifier
+// removed: if you run the above code in Ruby, you'll find that `bar` is not defined as a class method on `A`, but
+// rather as a not-always-available instance method on `A`, so introducing it as a static method is not at all
+// correct. Finally, because methods evaluate to their corresponding symbols, the former location of `bar`'s definition
+// has been replaced with the verbatim symbol `:bar`.
+//
+// It does this by maintaining both a stack and a queue: elements on the stack include indices into the queue as well as
+// scoping information, while the queue consists of pointers to expressions; additionally, some stack elements do not
+// correspond to expressions that need to be moved, in which case their indices will be -1: for example, all methods at
+// the root of a class definition will have stack entries (so we can properly track nested methods) but none of them
+// will need to be moved, so they will have -1 indices.
+namespace sorbet::dsl {
+
+class FlattenWalk {
+    struct MethodData {
+        // This is an index into the methods stack
+        int idx;
```

Comment:
what is this used for?

Reply:
Actually, I was too hasty in removing this, because we either need this index or something analogous to it in order to maintain the correct ordering. I'll rubber-duck about this in the channel.